### PR TITLE
create e2e test  for direct method 

### PR DESCRIPTION
**Pull request by dacolgit**
Create skeleton test for direct method for current publisher version and publisher 2.5.5

### CODE REVIEW

**Review by cristipogacean — APPROVED** [review]

File: e2e-tests/IIoTPlatform-E2E-Tests/Standalone/A_PublishSingleNodeStandaloneDirectMethodTestTheory.cs [hunk 1]
```diff
@@ -44,6 +47,9 @@ IIoTMultipleNodesTestContext context
         async Task SubscribeUnsubscribeDirectMethodTest(MessagingMode messagingMode) {
             var ioTHubEdgeBaseDeployment = new IoTHubEdgeBaseDeployment(_context);
             var ioTHubPublisherDeployment = new IoTHubPublisherDeployment(_context, messagingMode);
+            _iotHubConnectionString = _context.IoTHubConfig.IoTHubConnectionString;
+            _iotHubPublisherDeviceName = _context.DeviceConfig.DeviceId;
+            _iotHubPublisherModuleName = "publisher_standalone";
```
Comment (karok2m) [inline e2e-tests/IIoTPlatform-E2E-Tests/Standalone/A_PublishSingleNodeStandaloneDirectMethodTestTheory.cs]: Can you please move initialization of those class fields to `A_PublishSingleNodeStandaloneDirectMethodTestTheory()` constructor ? That way we can avoid repetition of this code in each test case.

Comment (dacolgit) [reply e2e-tests/IIoTPlatform-E2E-Tests/Standalone/A_PublishSingleNodeStandaloneDirectMethodTestTheory.cs]: I moved the initialization of `_iotHubConnectionString` and `_iotHubPublisherDeviceName` to the constructor.
The `_iotHubPublisherModuleName` depends on the test, if it is a test for current version of the publisher or the legacy version

File: e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs [hunk 2]
```diff
@@ -545,6 +545,112 @@ private static PrivateKeyFile GetPrivateSshKey(IIoTPlatformTestContext context)
...
+        /// Create a single node model with opcplc node
+        /// </summary>
+        /// <param name="IIoTMultipleNodesTestContext">context</param>
+        /// <param name="CancellationTokenSource">cancellation token</param>
+        public static async Task<PublishedNodesEntryModel> CreateSingleNodeModelAsync(IIoTMultipleNodesTestContext context, CancellationTokenSource cts) {
+            IDictionary<string, PublishedNodesEntryModel> simulatedPublishedNodesConfiguration =
+                new Dictionary<string, PublishedNodesEntryModel>(0);
+
+            // With the nested edge test servers don't have public IP addresses and cannot be accessed in this way
+            if (context.IoTEdgeConfig.NestedEdgeFlag != "Enable") {
```
Comment (karok2m) [inline e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs:558]: How relevant is the check for nested edge ? Do we also need it in `CreateMultipleNodesModelAsync()` method ?

File: e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs [hunk 3]
```diff
@@ -545,6 +545,112 @@ private static PrivateKeyFile GetPrivateSshKey(IIoTPlatformTestContext context)
...
+            return model;
+        }
+
+        /// <summary>
+        /// Create a multiple nodes model with opcplc nodes
+        /// </summary>
+        /// <param name="IIoTMultipleNodesTestContext">context</param>
+        /// <param name="CancellationTokenSource">cancellation token</param>
+        public static async Task<PublishedNodesEntryModel> CreateMultipleNodesModelAsync(IIoTMultipleNodesTestContext context, CancellationTokenSource cts) {
+            await context.LoadSimulatedPublishedNodes(cts.Token);
```
Comment (karok2m) [inline e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs]: Do we need to call this method in the beginning of `CreateSingleNodeModelAsync()` method as well ?

File: e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs [hunk 4]
```diff
@@ -545,6 +545,112 @@ private static PrivateKeyFile GetPrivateSshKey(IIoTPlatformTestContext context)
             return dictionary.ContainsKey(propertyName);
         }
 
+        /// <summary>
+        /// Create a single node model with opcplc node
+        /// </summary>
+        /// <param name="IIoTMultipleNodesTestContext">context</param>
+        /// <param name="CancellationTokenSource">cancellation token</param>
+        public static async Task<PublishedNodesEntryModel> CreateSingleNodeModelAsync(IIoTMultipleNodesTestContext context, CancellationTokenSource cts) {
```
Comment (karok2m) [inline e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs]: The method should be receiving `CancellationToken` object instead of `CancellationTokenSource`.

File: e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs [hunk 5]
```diff
@@ -545,6 +545,112 @@ private static PrivateKeyFile GetPrivateSshKey(IIoTPlatformTestContext context)
...
+
+            return model;
+        }
+
+        /// <summary>
+        /// Create a multiple nodes model with opcplc nodes
+        /// </summary>
+        /// <param name="IIoTMultipleNodesTestContext">context</param>
+        /// <param name="CancellationTokenSource">cancellation token</param>
+        public static async Task<PublishedNodesEntryModel> CreateMultipleNodesModelAsync(IIoTMultipleNodesTestContext context, CancellationTokenSource cts) {
```
Comment (karok2m) [inline e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs]: The method should be receiving `CancellationToken` object instead of `CancellationTokenSource`.

**Review by karok2m — APPROVED** [review]

File: e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs (hunk 3 above)
Comment (dacolgit) [reply e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs]: No this is just needed for multiple nodes. For single node it just publish a node with nodeId = `ns=2;s=SlowUInt1`

File: e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs (hunk 4 above)
Comment (dacolgit) [reply e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs]: changed

File: e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs (hunk 5 above)
Comment (dacolgit) [reply e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs]: changed

File: e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs (hunk 2 above)
Comment (dacolgit) [reply e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs:558]: Not sure this came from the original code

Comment (karok2m) [reply e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs:558]: OK, we can figure that out later. But we need to do testing in nested-edge setup as well at some point.

**Review by karok2m — APPROVED** [review]
